### PR TITLE
CIS-69 keystroke

### PR DESCRIPTION
**Pull request by buh**
Implementation of `keystroke` and `stopTyping` on LLC.

### CODE REVIEW

File: Sources/Client/Model/Channel/Channel+Requests.swift [hunk 1]
```diff
@@ -133,6 +135,37 @@ public extension Channel {
...
     
+    /// Send a keystroke event for the current user.
+    /// - Parameter completion: a completion block with `Event`.
+    @discardableResult
+    func keystroke(_ completion: @escaping Client.Completion<Event>) -> Cancellable {
+        currentUserTypingTimerControl?.cancel()
+        
+        currentUserTypingTimerControl = DefaultTimer.schedule(timeInterval: 5, queue: .main) { [weak self] in
+            self?.stopTyping(completion)
+        }
```
Comment (VojtaStavik) [inline Sources/Client/Model/Channel/Channel+Requests.swift]: It would be nice to extract the `5` into a preset which could be changed by a user.

File: Sources/Client/Model/Channel/Channel+Requests.swift [hunk 2]
```diff
@@ -133,6 +135,37 @@ public extension Channel {
...
+        
+        guard !isCurrentUserTyping else {
+            return Subscription.empty
+        }
```
Comment (VojtaStavik) [inline Sources/Client/Model/Channel/Channel+Requests.swift:150]: This feels iffy. If there are valid cases when calling `keystroke` doesn't produce a request, it needs to be reflected in the API:
- it should return `-> Cancellable?`
- `completion` should be called with a value indicating no request was actually made.

File: Sources/Client/Model/Channel/Channel+Requests.swift [hunk 3]
```diff
@@ -133,6 +135,37 @@ public extension Channel {
...
+        return Client.shared.send(eventType: .typingStart, to: self, completion)
+    }
+    
+    /// Send a stop typing event for the current user.
+    /// - Parameter completion: a completion block with `Event`.
+    @discardableResult
+    func stopTyping(_ completion: @escaping Client.Completion<Event>) -> Cancellable {
+        guard isCurrentUserTyping else {
+            return Subscription.empty
+        }
```
Comment (VojtaStavik) [inline Sources/Client/Model/Channel/Channel+Requests.swift:163]: ... same here

File: Sources/Client/Model/Channel/Channel.swift [hunk 4]
```diff
@@ -117,6 +117,9 @@ public final class Channel: Codable {
     
     private var subscriptionBag = SubscriptionBag()
     
+    var isCurrentUserTyping = false
```
Comment (VojtaStavik) [inline Sources/Client/Model/Channel/Channel.swift:120]: This needs to be `atomic`

File: Sources/Client/Model/Channel/Channel.swift [hunk 5]
```diff
@@ -117,6 +117,9 @@ public final class Channel: Codable {
...
+    var currentUserTypingTimerControl: TimerControl?
```
Comment (VojtaStavik) [inline Sources/Client/Model/Channel/Channel.swift:121]: This should be `atomic` too

File: Sources/Core/Presenter/Channel/ChannelPresenter+Events.swift [hunk 6]
```diff
@@ -28,15 +28,13 @@ extension ChannelPresenter {
...
+            guard isTypingEventsEnabled else {
                 return .none
             }
             
             let shouldUpdate = filterInvalidatedTypingUsers()
             
-            if channel.config.typingEventsEnabled,
-                !user.isCurrent,
-                (typingUsers.isEmpty || !typingUsers.contains(.init(user: user))) {
+            if !user.isCurrent, (typingUsers.isEmpty || !typingUsers.contains(.init(user: user))) {
```
Comment (VojtaStavik) [inline Sources/Core/Presenter/Channel/ChannelPresenter+Events.swift:37]: This `if` statement is hard to parse "visually". It would be nice to introduce some local variables to make it easier for your teammates to understand.

File: Sources/Client/Model/Channel/Channel+Requests.swift [hunk 7]
```diff
@@ -133,6 +135,37 @@ public extension Channel {
...
+        
+        currentUserTypingTimerControl = DefaultTimer.schedule(timeInterval: 5, queue: .main) { [weak self] in
+            self?.stopTyping(completion)
+        }
+        
+        guard !isCurrentUserTyping else {
+            return Subscription.empty
+        }
+        
+        isCurrentUserTyping = true
```
Comment (VojtaStavik) [inline Sources/Client/Model/Channel/Channel+Requests.swift:152]: Maybe we should assign this after the request has succeeded? Or it can stay like this bu we should reset it if the request fails.

File: Sources/Client/Model/Channel/Channel+Requests.swift (hunk 2 above)
Comment (VojtaStavik) [reply Sources/Client/Model/Channel/Channel+Requests.swift:150]: ... or at least add to the documentation comment that `completion` may never get called if `isCurrentUserTyping` is already `true`.

File: Sources/Client/Model/Channel/Channel+Requests.swift (hunk 1 above)
Comment (buh) [reply Sources/Client/Model/Channel/Channel+Requests.swift]: Let's leave it for now with a value from the task description.

File: Sources/Client/Model/Channel/Channel.swift (hunk 4 above)
Comment (buh) [reply Sources/Client/Model/Channel/Channel.swift:120]: It's should be always called from the main thread.

File: Sources/Client/Model/Channel/Channel.swift (hunk 5 above)
Comment (buh) [reply Sources/Client/Model/Channel/Channel.swift:121]: The same here... only main thread.

Comment (buh) [reply Sources/Client/Model/Channel/Channel.swift:121]: I'll add comments.

File: Sources/Core/Presenter/Channel/ChannelPresenter+Events.swift (hunk 6 above)
Comment (buh) [reply Sources/Core/Presenter/Channel/ChannelPresenter+Events.swift:37]: I agree. But here, is it unclear for you?

File: Sources/Client/Model/Channel/Channel.swift (hunk 4 above)
Comment (VojtaStavik) [reply Sources/Client/Model/Channel/Channel.swift:120]: I don't see a reason why this function must be called from the main thread tbh.

File: CHANGELOG.md [hunk 8]
```diff
@@ -6,6 +6,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 # Upcoming
 
+### ✅ Added
+- Send a keystroke event for the current user: `channel.keystroke {}`. You need to call it each time when the current user press a key. The method will automatically send a typing stop event after 15 seconds timeout after a late call of `keystroke()`.
```
Comment (VojtaStavik) [inline CHANGELOG.md:10]: `The method will automatically send a typing stop event after 15 seconds after the last call of `keystroke()`.`

File: CHANGELOG.md [hunk 9]
```diff
@@ -6,6 +6,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+- Send a stop typing event for the current user: `stopTyping {}`. Usually you no need to call it directly.
```
Comment (VojtaStavik) [inline CHANGELOG.md:11]: `Usually, you don't need to call this method directly.`
